modules/ml/sagemaker-custom-kernel/scripts/create_kernel.py

```diff
@@ -0,0 +1,210 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+import json
+import os
+
+import boto3
+
+# Seed-Farmer Variables
+project_name = os.getenv("SEEDFARMER_PROJECT_NAME", "")
+deployment_name = os.getenv("SEEDFARMER_DEPLOYMENT_NAME", "")
+module_name = os.getenv("SEEDFARMER_MODULE_NAME", "")
+
+# Module Parameters
+custom_kernel_name = os.getenv("SEEDFARMER_PARAMETER_CUSTOM_KERNEL_NAME", "")
+image_name = os.getenv("SEEDFARMER_PARAMETER_SAGEMAKER_IMAGE_NAME")
+app_image_config_name = os.getenv("SEEDFARMER_PARAMETER_APP_IMAGE_CONFIG_NAME")
+sm_studio_domain_id = os.environ.get("SEEDFARMER_PARAMETER_STUDIO_DOMAIN_ID")
+sm_studio_domain_name = os.environ.get("SEEDFARMER_PARAMETER_STUDIO_DOMAIN_NAME")
+uid = os.getenv("SEEDFARMER_PARAMETER_KERNEL_USER_UID", 1000)
+gid = os.getenv("SEEDFARMER_PARAMETER_KERNEL_USER_GID", 100)
+mount_path = os.getenv("SEEDFARMER_PARAMETER_KERNEL_USER_HOME_MOUNT_PATH", "/home/sagemaker-user")
+
+# CDK Export
+cdk_exports_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), "..", "cdk-exports.json")
+cdk_output_file = open(cdk_exports_path)
+cdk_output = json.load(cdk_output_file)
+app_prefix = f"{project_name}-{deployment_name}-{module_name}"
+role_arn = json.loads(cdk_output[app_prefix]["metadata"])["SageMakerCustomKernelRoleArn"]
+
+# Deployspec ENV vars
+ecr_repository_uri = os.getenv("REPOSITORY_URI")
+image_uri = os.getenv("IMAGE_URI")
+
+# Config vars
+display_name = image_name  # Should stay same to allow config mapping in update_domain
+sm_client = boto3.client("sagemaker")
+tags = [
+    {"Key": "CreatedBy", "Value": "IDF"},
+    {"Key": "ProjectName", "Value": project_name},
+    {"Key": "ModuleName", "Value": module_name},
+    {"Key": "DeploymentName", "Value": deployment_name},
+]
+
+
+def create_image():
+    return sm_client.create_image(
+        Description="IDF Generated Image",
+        DisplayName=display_name,
+        ImageName=image_name,
+        RoleArn=role_arn,
+        Tags=tags,
+    )
+
+
+def create_image_if_not_exist():
+    print("Attempting to create Image:", image_name)
+    try:
+        image = create_image()
+        print("Image Created:", image)
+    except Exception as e:
+        print(f"Warning: already exists [error: {e}]. Skipping creation...")
+
+
+def create_image_version():
+    sm_client.create_image_version(
+        BaseImage=image_uri,
+        ImageName=image_name,
+    )
+
+    return sm_client.describe_image_version(ImageName=image_name)
+
+
+def check_image_version_exists():
```

Comment:
I am not an SM expert, but is there a CDK way of doing this, like using an L3 construct?

Reply:
There [is](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_sagemaker.CfnImageVersion.html) an L1 construct for image version, but I will need to check whether it does everything we need it to do.

Reply:
Refactored to use CDK constructs where possible - `Image`, `ImageVersion`, and `AppImageConfig` - check the latest commit please. Dealing with L1 is a bit annoying but arguably still easier than maintaining a custom script.